### PR TITLE
[SYCLomatic] Enable migration of __shfl* function with mask

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

**Review by ziranzha — APPROVED** [review]

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 1]
```diff
@@ -15,6 +15,31 @@
 #include <cassert>
 #include <cstdint>
 
+// TODO: Remove these function definitions once they exist in the DPC++ compiler
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp:18]: can those API works with computecpp? If not, we may need add macro to protect them

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 2]
```diff
@@ -384,6 +409,208 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+      g.get_local_linear_id() / logical_sub_group_size * logical_sub_group_size;
+  unsigned logical_remote_id =
+      start_index + remote_local_id % logical_sub_group_size;
+#if defined(__SYCL_DEVICE_ONLY__)
+#if defined(__SPIR__)
+  return detail::masked_sub_group_call(member_mask, g, [=]() {
+        return __spirv_GroupNonUniformShuffle(__spv::Scope::Subgroup, x, logical_remote_id);
+    });
+#elif defined(__NVPTX__)
+   // TODO: May need to call this twice for 64-bit types
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: remove the TODO or fix it.

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 3]
```diff
@@ -384,6 +409,208 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+   return __nvvm_shfl_sync_idx_i32(mask, x, logical_remote_id, 0x1f);
+#else
+  #error "select_from_sub_group only supports SPIR-V and NVPTX backends"
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: Masked version of select_from_sub_group

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 4]
```diff
@@ -384,6 +409,208 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+#else
+  #error "shift_sub_group_left only supports SPIR-V and NVPTX backends"
+#endif // __SPIR__
+#else
+  (void)g;
+  (void)x;
+  (void)delta;
+  (void)logical_sub_group_size;
+  (void)member_mask;
+  throw sycl::exception(sycl::errc::runtime, "Not supported on host device.");
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: API...   not supported on host side.

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 5]
```diff
@@ -384,6 +409,208 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+#if defined(__SPIR__)
+  T result = detail::masked_sub_group_call(member_mask, g, [=]() {
+    return __spirv_GroupNonUniformShuffleUp(__spv::Scope::Subgroup, x, delta);
+  });
+  if ((id - start_index) < delta) {
+    result = x;
+  }
+  return result;
+#elif defined(__NVPTX__)
+   // TODO: May need to call this twice for 64-bit types
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: need fix the todo.

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 6]
```diff
@@ -372,6 +372,9 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         false),
              DPCT_OPT_ENUM("user-defined-reductions", int(ExperimentalFeatures::Exp_UserDefineReductions),
         "Experimental extension that allows user define reductions.\n",
+        false),
+             DPCT_OPT_ENUM("masked_sub_group_function", int(ExperimentalFeatures::Exp_MaskedSubGroupFunction),
+        "Experimental maksed sub_group call.\n",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: Please ref to other option in --use-experimental-features for the description.

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 1 above)
Comment (intwanghao) [reply clang/test/dpct/helper_files_ref/include/util.hpp:18]: can not work with computecpp

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 7]
```diff
@@ -374,7 +374,7 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         "Experimental extension that allows user define reductions.\n",
         false),
              DPCT_OPT_ENUM("masked_sub_group_function", int(ExperimentalFeatures::Exp_MaskedSubGroupFunction),
-        "Experimental maksed sub_group call.\n",
+        "Experimental helper function used to help maksed sub_group call during migration.\n",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: masked 
```suggestion
        "Experimental extension that allows masked sub_group call during migration.\n",
```

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 8]
```diff
@@ -374,7 +374,7 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         "Experimental extension that allows user define reductions.\n",
         false),
              DPCT_OPT_ENUM("masked_sub_group_function", int(ExperimentalFeatures::Exp_MaskedSubGroupFunction),
-        "Experimental helper function used to help maksed sub_group call during migration.\n",
+        "Experimental extension that allows masked sub_group call during migration.\n",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc:377]: As followup, please create a doc JIRA.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**